### PR TITLE
Eventnotification

### PR DESCRIPTION
**Pull request by somvitbhowmik**
Generating a new pull request to merge with develop branch.

### CODE REVIEW

**Comment by egov-systems** [comment]
Please take care of the following review points, we are keeping this pull open till you fix the below points and commit the files,  we will do the remaining review after these are fixed. Also ensure to check across your module for similar issues, so that we don't have to report it again.

1) If **UserFcmDevice** is part of different module then rename the table same as module name (eg: instead of **egevntnotification_userfcmdevice** it must be **egpushbox_userfcmdevice**) and change corresponding sequence name as well.

2) In UserFcmDevice replace 
`@Unique(id = "id", tableName = "egevntnotification_userfcmdevice", columnName = { "userid", "devicetoken", "deviceid" }, fields = { "userId", "userDeviceToken", "deviceid" }, enableDfltMsg = true)`
with 
`@Unique(fields = { "userId", "userDeviceToken", "deviceid" }, enableDfltMsg = true)`

3) Did not understand why **@ManyToOne** and **@Embedded**  together in UserFcmDevice#User, please clarify.

4) In **UserFcmDevice**, remove **@Column(name = "devicetoken", nullable = false)** from **userDeviceToken** and either rename the corresponding column same as the attribute's name or rename field same to the column name. If you want this field to be non nullable then ensure to put **@NotBlank** annotation

5) Ensure to put appropriate/missing jpa validation annotation for all entities (eg: **UserFcmDevice#userDeviceToken** missing **@NotBlank**)

6) Rename **MessageContent#messageContentDetails** field to **details**

7)  Some of the class names are not making sense, if possible rename such classes to an understandable way eg : **SendMessageList** . Unable to understand, either rename it properly or add some javadocs to understand the usage of the class (Ensure to do the same across your new class files). Also add proper documentation at complex method level to understand its functionality (No need to add javadoc for getters/setters or for any simple methods, add it only for complex and difficult to understand kind of methods)

8)  Move **config** package outside the **entity** package

9) In **PushNotificationService** Don't Autowire `PushboxConfiguration pushboxConfiguration` its spring config class for property bean you have to autowire `PushboxProperties`

10) No exception should be ignored, it must be logged and rethrown (eg: **PushNotificationService#setUpFirebaseApp()** and **sendMessagesToDevices**)

11) **PushNotificationService#saveUserDevice**  `if (null != existingRecord)` reverse this to `if (existingRecord != null) `though both are effectively same, the second have more readability.

**Comment by egov-systems** [comment]
1) **deviceId** in **UserFcmDevice** can be blank ? If not add @NotBlank, also as i have mentioned in my previous review, please cross check across your changes and add the missing validation annoation if any.

2)In **UserFcmDevice** field **user** should be anotated with @NotNull (@NotBlank applicable only for String fields)

3) Did not understand why values are getting set after retriving from DB at code https://github.com/egovernments/eGov/blob/b775387d893cbab734b93b8b4b8eae2c2d48587f/egov/egov-pushbox/src/main/java/org/egov/pushbox/service/PushNotificationService.java#L98 to https://github.com/egovernments/eGov/blob/b775387d893cbab734b93b8b4b8eae2c2d48587f/egov/egov-pushbox/src/main/java/org/egov/pushbox/service/PushNotificationService.java#L100

4) The DB migration script vesrion must be in order, now core DDL version is latest than DML. So rename the script in order of execution. Only 2 migration script is enough, one is for Core DDL and other one is for Core (Seed) data.

5) No need of drop scripts for the tables and sequence, etc. Recreate database and run the migration  scripts ( as per point 4) from scratch.

6) This value seems wrong https://github.com/egovernments/eGov/blob/b775387d893cbab734b93b8b4b8eae2c2d48587f/egov/egov-eventnotification/src/main/resources/config/eventnotification-application-config.properties#L63

**Comment by egov-systems** [comment]
1) Previous review point no 6 is not correct yet, key is `eventnotification.propertytax.rest` for property tax and value of that key is water charges ? and also if its a URL then why its having space in the url part ? /wtms/rest/watertax/**Water Tax Management** 

2) https://github.com/egovernments/eGov/blob/eventnotification/egov/egov-pushbox/src/main/resources/db/migration/main/V20180622130018__egpushbox_core_ddl.sql and https://github.com/egovernments/eGov/blob/eventnotification/egov/egov-eventnotification/src/main/resources/db/migration/main/V20180626142847__egeventnotification_core_ddl.sql still contains drop script

3)  Don't Autowire custom repository https://github.com/egovernments/eGov/blob/913d488131490ea92dcb0a649371d0f905e02547/egov/egov-eventnotification/src/main/java/org/egov/eventnotification/service/DraftService.java#L67, check how spring data jpa custom repository is used from any existing module or spring reference doc (correct the similar issues across)

4)  Not threadsafe since this job instance is created as singleton bean https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotification/src/main/java/org/egov/eventnotification/scheduler/NotificationSchedulerJob.java#L110 and https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotification/src/main/java/org/egov/eventnotification/scheduler/NotificationSchedulerJob.java#L111

5) Rename all the jsp and javascript, css files to small case and delimit with hyphen (-) same as https://github.com/egovernments/eGov/blob/eventnotification/egov/egov-eventnotificationweb/src/main/webapp/resources/js/app/schedule-update-view.js

6) Remove suffix/prefix like **helper** and **common** from all resource files and name it appropriately 

7) Remove **decorators** and **includes** folders from https://github.com/egovernments/eGov/tree/eventnotification/egov/egov-eventnotificationweb/src/main/webapp

8) Remove **classes** folder and its content https://github.com/egovernments/eGov/tree/eventnotification/egov/egov-eventnotificationweb/src/main/webapp/WEB-INF/

9) Correct     https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotificationweb/src/main/webapp/WEB-INF/web.xml#L54

10) Remove all unused tags and taglibs from https://github.com/egovernments/eGov/tree/eventnotification/egov/egov-eventnotificationweb/src/main/webapp/WEB-INF/tags and https://github.com/egovernments/eGov/tree/eventnotification/egov/egov-eventnotificationweb/src/main/webapp/WEB-INF/taglib

11) Remove https://github.com/egovernments/eGov/blob/eventnotification/egov/egov-eventnotificationweb/src/main/webapp/WEB-INF/views/template/dashboard.jsp

12) Remove setting Id back at https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotificationweb/src/main/java/org/egov/eventnotification/web/controller/event/ModifyEventController.java#L122

13) Move code from line no https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotificationweb/src/main/java/org/egov/eventnotification/web/controller/event/ModifyEventController.java#L123 to https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotificationweb/src/main/java/org/egov/eventnotification/web/controller/event/ModifyEventController.java#L133 to its service class

14) https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotificationweb/src/main/java/org/egov/eventnotification/web/controller/event/ModifyEventController.java#L105 , Arrays.asList returns a List then why again initializing it into ArrayList

15) Rename this URL as per convention https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotificationweb/src/main/java/org/egov/eventnotification/web/controller/notificationdraft/NotificationDraftController.java#L112

16) Move line from https://github.com/egovernments/eGov/blob/eventnotification/egov/egov-eventnotificationweb/src/main/java/org/egov/eventnotification/web/controller/schedule/ModifyScheduleController.java#L146 to https://github.com/egovernments/eGov/blob/eventnotification/egov/egov-eventnotificationweb/src/main/java/org/egov/eventnotification/web/controller/schedule/ModifyScheduleController.java#L187 to service class

17) Move line from https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotificationweb/src/main/java/org/egov/eventnotification/web/controller/schedule/ScheduleController.java#L174 to https://github.com/egovernments/eGov/blob/33d83de0fe7f5629da3231c4e0b511fead34083f/egov/egov-eventnotificationweb/src/main/java/org/egov/eventnotification/web/controller/schedule/ScheduleController.java#L207 to service class

18) Move non view related code from controller to its corresponding services classes wherever applicable.

@satyamashish please ensure this module is developed as per the requirement & design and approve this pull request if everything is checked and verified. Also please get the approval of the person who tested this module in test environment before merging.
@amit-panwar and @jayashree2015 please approve this pull from your side if the developer followed and solved all the review points raised by you.
@vasanth-egov please approve if everything is as per the standard for merge.

@satyamashish  Once the above review points are covered and above people gives approval then we can proceed with the merge. Please ensure that technical documentation as well as KT to a SE is completed from the developer.

**Comment by somvitbhowmik** [comment]
As mentioned point no 1. That URL I got from the developer who has developed the water tax rest API. Once he change the URL I will update in the properties file. Rest of the point I will validate and take care.

**Comment by egov-systems** [comment]
@shashikanth12345 have you given property tax REST api url like `/wtms/rest/watertax/ Water Tax Management` ? now as per @somvitbhowmik the value he is using for both property tax and water charges  rest api url is same.

**Comment by egov-systems** [comment]
@somvitbhowmik https://github.com/egovernments/eGov/pull/37#issuecomment-399437729, Point No. 3 not yet clarified.

**Comment by egov-systems** [comment]
https://github.com/egovernments/eGov/pull/37#issuecomment-400570767 Did not understand, why If already exist then to override the same value ? I don't think neither User nor deviceId changes once its stored.

**Comment by egov-systems** [comment]
1) Format the pom, jsp, js, java files once again (hope you have already imported, format std xml to your eclipse IDE from our repository, if not get in touch with @amit-panwar ).

2) Remove https://github.com/egovernments/eGov/blob/a3fadd6cb005b8cfe9be6559ab84814e5755eaca/egov/egov-pushbox/src/main/java/org/egov/pushbox/entity/UserFcmDevice.java#L97. **@Column** is only required if field name and column name is different or we want to auto generate tables (Do the same across all entities).

3) For **UserFcmDevice** if any composite uniqueness to be checked then use **@CompositeUnique** instead of **@Unique**

4)  Add ` if(LOGGER.infoEnabled())` before https://github.com/egovernments/eGov/blob/ae6e11c42df2343629f6b57feadfecf96005fd4c/egov/egov-pushbox/src/main/java/org/egov/pushbox/service/PushNotificationService.java#L134 and do the same across source wherever logging the information or debug

5) Is this a one time activity or every time we have to call this initialize function ? https://github.com/egovernments/eGov/blob/ae6e11c42df2343629f6b57feadfecf96005fd4c/egov/egov-pushbox/src/main/java/org/egov/pushbox/service/PushNotificationService.java#L150 , If its one time activity then ensure to add it in a @PostConstruct annotated method to initialize and remove direct call from https://github.com/egovernments/eGov/blob/ae6e11c42df2343629f6b57feadfecf96005fd4c/egov/egov-pushbox/src/main/java/org/egov/pushbox/service/PushNotificationService.java#L133 .

6) Use slf4j logger everywhere and replace https://github.com/egovernments/eGov/blob/ae6e11c42df2343629f6b57feadfecf96005fd4c/egov/egov-pushbox/src/main/java/org/egov/pushbox/service/PushNotificationService.java#L193 with `LOGGER.info("##PushBoxFox## : Message Send Status : {}", response)`. Also do the same wherever applicable.

7) Replace https://github.com/egovernments/eGov/blob/ae6e11c42df2343629f6b57feadfecf96005fd4c/egov/egov-pushbox/src/main/java/org/egov/pushbox/service/PushNotificationService.java#L195 with `LOGGER.error("##PushBoxFox## : Error : Encountered an exception while sending the message ", ex)`; and do the same wherever logging error.

File: egov/egov-api/src/main/java/org/egov/api/controller/RestPushBoxController.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    @Autowired
+    private PushNotificationService notificationService;
+
+    @PostMapping(path = UPDATE_USER_TOKEN, consumes = APPLICATION_JSON_VALUE)
+    public @ResponseBody ResponseEntity<String> updateToken(@RequestBody UserTokenRequest userTokenRequest) {
+        ApiResponse res = ApiResponse.newInstance();
+        try {
+
+            if (isBlank(userTokenRequest.getUserToken()))
+                return res.error(getMessage("userdevice.device.tokenunavailable"));
```
Comment (amit-panwar) [inline egov/egov-api/src/main/java/org/egov/api/controller/RestPushBoxController.java]: Please use validator to validate the request object .

File: egov/egov-api/src/main/java/org/egov/api/controller/RestEventController.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    @GetMapping(path = "/event/search", produces = APPLICATION_JSON_VALUE)
+    public ResponseEntity<String> searchEvent(EventSearch eventSearch) {
+
+        try {
+            return getResponseHandler()
+                    .setDataAdapter(new EventSearchAdapter(userEventService))
+                    .success(eventService.searchEvent(eventSearch));
+        } catch (final Exception e) {
```
Comment (amit-panwar) [inline egov/egov-api/src/main/java/org/egov/api/controller/RestEventController.java]: Please use exception handler to catch the exception and remove try catch block.

**Comment by egov-systems** [comment]
@somvitbhowmik please do an end to end review on all codes added or updated by you with the guidelines provided by us and the codacy at various review process, once you are fully done with all the fixes then please create a new pull request (Before creating the next pull request, please send us a mail we will do the final review on the branch itself and confirm you for a pull request).

**Pull request closed without merging**